### PR TITLE
DGR-570: Update moodle CI with current supported versions

### PR DESCRIPTION
**Pull request by kei178**
This PR upgrades the development Moodle version to v4.3 and also updates the CI environments with recent versions.

The combinations of supported Moodle, PHP, and DB versions can be found in the official documents below:

- https://download.moodle.org/releases/latest/
- https://download.moodle.org/releases/supported/
- https://download.moodle.org/releases/legacy/

I've made sure all CI checks pass from Moodle 3.11 to 4.4 and all credential issuance features (manual issuance, auto-issuance based on quiz completion and course completion) work as expected in the Moodle 4.3 docker-compose setup.

Following [this example GA yml](https://github.com/moodlehq/moodle-plugin-ci/blob/cdb52da8f712f9ab7dd558378236afb2979a7c17/gha.dist.yml#L55), I also tried to upgrade the version of `moodle-plugin-ci` from 3 to 4, but there were too many code checker violations so I gave up halfway through and decided to address it some other time.

That's why you can see a lot of changes to fix "Short array syntax must be used to define arrays" (`array() -> []`) and "There should be a comma after the last array item in a multi-line array." (see [this job failures](https://github.com/accredible/moodle-mod_accredible/actions/runs/9405654734/job/25907337812) as an example), but essentially this PR is updating `.github/workflows/moodle-ci.yml`, `Dockerfile` and `docker-compose.yml`.

---

Putting `not-ready-for-merge` just because waiting for the dynamic attribute mapping PR from @Ndibui

### CODE REVIEW

File: .github/workflows/moodle-ci.yml [hunk 1]
```diff
@@ -31,20 +31,26 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        php: ['7.3', '7.4']
-        moodle-branch: ['MOODLE_310_STABLE', 'MOODLE_311_STABLE']
+        php: ['8.0']
+        moodle-branch: ['MOODLE_401_STABLE', 'MOODLE_402_STABLE', 'MOODLE_403_STABLE']
         database: [pgsql, mariadb]
         include:
```
Comment (albertpaulp) [inline .github/workflows/moodle-ci.yml:37]: Why are we testing different Moodle branches with both Postgres and MariaDB? Is that recommended that all integrations is expected to work with both DBs?

**Review by albertpaulp — APPROVED** [review]
I guess the major differences affecting our codebase between the PHP versions are just Array syntax?

**Review by shclarke007 — APPROVED** [review]

**Review by abhisheksarka — APPROVED** [review]

File: .github/workflows/moodle-ci.yml (hunk 1 above)
Comment (Ndibui) [reply .github/workflows/moodle-ci.yml:37]: Moodle versions expect specific versions of PHP to work. We have to define the matrix like this to ensure the  Moodle versions match their PHP versions. Ideally, we ought to test the last 3 major versions.

**Review by Ndibui — APPROVED** [review]

File: .github/workflows/moodle-ci.yml (hunk 1 above)
Comment (kei178) [reply .github/workflows/moodle-ci.yml:37]: We don't have control over which version and database our clients are using hence it is recommended to use multiple setups in the CI tests.

**Comment by kei178** [comment]
> I assume the major differences affecting our codebase between the PHP versions are just Array syntax?

@albertpaulp - The array syntax violation that I mentioned is due to Moodle Code Checker version change, not the PHP version.

I looked through the [breaking changes between 7.4.x and 8.0.x](https://www.php.net/manual/en/migration80.incompatible.php) and I think there is nothing to address for this plugin.

Having said that, the unit test coverage is limited and thanks to your comment, I've discovered our docker dev env is using Moodle 4.3.4 with PHP 7.4.30 so I think I should upgrade the PHP version to 8.0 and manually make sure everything works before merging this.

**Comment by albertpaulp** [comment]
@kei178 
35xh5n

**Comment by kei178** [comment]
I upgraded the dev PHP version to 8, manually tested all credential issuance scenarios, and confirmed our plugin works in PHP 8 too.

I saw only the following warning so I fixed it with this change: https://github.com/accredible/moodle-mod_accredible/pull/107/commits/982974dd42751b359a3d1481fe7d253791ff58d5

```
Deprecated: rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /bitnami/moodle/mod/accredible/classes/apirest/apirest.php on line 83
```

**Pull request merged**